1024 add a collective broadcast call for collections

JacobDomagala opened the pull request:
What needs to be done:

1. Add colective bcast for collection:
- [x] with member msg hanldler
- [x] with non-member msg hanldler
- [x] lb and trace related calls
2. Add unittests:
- [x] for member handler version
- [x] for non-member handler version

3. Add documentation:
- [x]  documentation

Fixes #1024

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_collection_group.extended.cc  4
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

JacobDomagala commented on src/vt/vrt/collection/manager.impl.h:
```diff
@@ -848,6 +848,71 @@ messaging::PendingSend CollectionManager::broadcastFromRoot(MsgT* raw_msg) {
+
+  return messaging::PendingSend(
+    promoMsg, [proxy](MsgSharedPtr<BaseMsgType>& msgIn) {
+      auto elm_holder = theCollection()->findElmHolder<ColT, IndexT>(proxy);
```
I think it's just easier to call collectionBcastHandler() here. Will try it out.

JacobDomagala commented on tests/unit/collection/test_collection_group.extended.cc:
```diff
@@ -96,4 +118,53 @@ TEST_F(TestCollectionGroup, test_collection_group_1) {
+  proxy.broadcastCollective<ColA::TestDataMsg, colHanlder>(msg.get());
+  EXPECT_EQ(elem_counter, 0);
+
+  proxy.broadcastCollective<ColA::TestDataMsg, &ColA::memberHanlder>(msg);
```
Should be non-member handler, i.e. `colHanlder`

JacobDomagala commented on tests/unit/collection/test_collection_group.extended.cc:
```diff
@@ -96,4 +118,53 @@ TEST_F(TestCollectionGroup, test_collection_group_1) {
+  EXPECT_EQ(elem_counter, -numElems);
+
+  proxy.broadcastCollective<
+    ColA::TestDataMsg, &ColA::memberHanlder, ColA::TestDataMsg>(my_node);
```
same here

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  14
- tests/unit/collection/test_collection_group.extended.cc  2
- src/vt/vrt/collection/manager.impl.h  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  14
- tests/unit/collection/test_collection_group.extended.cc  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

PhilMiller commented on src/vt/vrt/collection/manager.impl.h:
```diff
@@ -848,6 +848,81 @@ messaging::PendingSend CollectionManager::broadcastFromRoot(MsgT* raw_msg) {
+
+#if vt_check_enabled(lblite)
+  msg->setLBLiteInstrument(instrument);
+  msg->setCat(balance::CommCategory::NodeToCollection);
```
This is going to be a little bit of a miscue if we actually try to use the communication data in load balancing. It will look like every object communicates with the node currently hosting it, even though migrating it elsewhere won't cause any additional network communication. That makes me think we should find a way to just not count the communication aspect of these messages.

lifflander replied:
That a good point. Maybe we should add a new type of communication edge for collective broadcasts?

lifflander replied:
@JacobDomagala Can you add a new `CommCategory` for collective node to collection edges?

JacobDomagala replied:
Sure

JacobDomagala replied:
Out of curiosity, we want to add new communication edge to treat this type of communication in a unique way, or we want to ignore it entirely? For the latter, I think we could just set the 'lb_lite_instrument_' for this message to false. I don't know if this would have any negative effects, but from what I've seen it's mostly used to record the communication aspect (and also phase timing which I guess could be useful)

PhilMiller replied:
We definitely want the time spent to be instrumented, so setting that to false is not an option.

The communication here is never going to cross a network or process boundary. I see no reason to actually count it in that regard.

lifflander replied:
@PhilMiller I agree that we definitely need to instrument the call's work. We could skip the CommCategory if we don't see any use for LB to know that it happened as part of a communication. It might be asymmetric if all other invocations have a communication arrow (even local ones) and this one doesn't. It is true that it will never become non-local even after migrating to another node.

JacobDomagala replied:
Ok thanks for clearing that out

lifflander commented on src/vt/vrt/collection/manager.h:
```diff
@@ -904,6 +904,55 @@ struct CollectionManager
+   */
+  template <
+    typename MsgT,
+    ActiveColTypedFnType<MsgT,typename MsgT::CollectionType> *f>
```
On the indentation here, can you wrap the last `>` and indent as shown below?

```C++
template <
  typename MsgT,
  ActiveColTypedFnType<MsgT,typename MsgT::CollectionType> *f
>
messaging::PendingSend broadcastMsgCollective(
  CollectionProxyWrapType<typename MsgT::CollectionType> const& proxy,
  MsgT* msg, bool instrument = true
);
```

lifflander commented on src/vt/vrt/collection/manager.impl.h:
```diff
@@ -848,6 +848,81 @@ messaging::PendingSend CollectionManager::broadcastFromRoot(MsgT* raw_msg) {
+  using ColT = typename MsgT::CollectionType;
+
+  msg->setVrtHandler(
+    auto_registry::makeAutoHandlerCollectionMem<ColT, MsgT, f>());
```
Fix indentation here. Put last parenthesis on a new line.

lifflander commented on src/vt/vrt/collection/manager.impl.h:
```diff
@@ -848,6 +848,81 @@ messaging::PendingSend CollectionManager::broadcastFromRoot(MsgT* raw_msg) {
+}
+
+template <typename MsgT, typename ColT>
+messaging::PendingSend CollectionManager::broadcastMsgCollectiveImpl(
```
I think this needs to go through the scheduler instead of dispatching right away.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_collection_group.extended.cc  2
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

pnstickne commented on src/vt/vrt/collection/broadcast/broadcastable.h:
```diff
@@ -79,6 +79,24 @@ struct Broadcastable : BaseProxyT {
+  template <typename MsgT, ActiveColTypedFnType<MsgT, ColT> *f>
+  messaging::PendingSend broadcastCollective(MsgT* msg) const;
+  template <typename MsgT, ActiveColTypedFnType<MsgT, ColT> *f>
+  messaging::PendingSend broadcastCollective(MsgSharedPtr<MsgT> msg) const;
```
I've merged in https://github.com/DARMA-tasking/vt/pull/1028. This allows usage of MsgThief (see `Message::send`) and can now be used on "sending/ownership" transfer requests.. like sending. This will also eliminate one of these overheads.

pnstickne commented on src/vt/vrt/collection/broadcast/broadcastable.h:
```diff
@@ -79,6 +79,24 @@ struct Broadcastable : BaseProxyT {
   >
   messaging::PendingSend broadcast(Args&&... args) const;
+
+  template <typename MsgT, ActiveColTypedFnType<MsgT, ColT> *f>
```
Separate out the the construction and non-construction forms.

I recommend `broadcastCollectionMsg` when taking a MsgPtr (or MsgThief..) and using `broadcastCollective` when construction via forwarding.

JacobDomagala replied:
I did it this way to be consistent with the implementation of (rooted) broadcast present in this header. 

With the changes you propose i guess we should also change the `broadcast()` to perhaps `broadcastRooted`? And also we could apply the change with `MsgThief` which would let us "combine" both **Msg*** and **SmartPtr<Msg>** calls.

lifflander replied:
I don't want to rename `broadcast` at this point. That would break a lot of code. In the future, it's a possibility.

Also, I agree to be consistent with `broadcast` the message variant should just be an overload. I want this API to be symmetric with the regular `broadcast`. In the future, we might rename/refactor, but that would be a breaking change.

pnstickne replied:
For a _forward_ migration path:

1) New methods added that don't forward `broadcastMsg` (also, only one is needed..)

2) For the existing `broadcast`-with-overload, add `broadcastMsg` for uniformity and deprecated the existing `broadcast(msg)`.

lifflander replied:
@PhilMiller Comments on this?

lifflander replied:
I am generally in favor to making the split at some point. I'm just being cautious because of all the issues we've had with breaking APIs across release/develop in EMPIRE and aware of the amount of work that has been created to support both simultaneously.

I think switching to deprecated sounds fine to me. I would like other's opinions on this also.

@PhilMiller @nlslatt

nlslatt replied:
If we can swap between vt release and develop within EMPIRE without having to change EMPIRE code, I'm fine with it.  Otherwise, we should plan to change both branches simultaneously.

lifflander replied:
Please see comment below.

pnstickne commented on src/vt/vrt/collection/broadcast/broadcastable.h:
```diff
@@ -79,6 +79,24 @@ struct Broadcastable : BaseProxyT {
+  >
+  messaging::PendingSend broadcastCollective(Args&&... args) const;
+
+  template <typename MsgT, ActiveColMemberTypedFnType<MsgT, ColT> f>
```
So at the end, there should be 2 new method groups (different names) with 2 different templates each - only 4 methods total.

lifflander replied:
@pnstickne I think this is something we can consider for the future. But, for now, I don't want to break/change this API. Broadcast calls are prevalent across EMPIRE and many examples and changing it would be make it nearly impossible to keep supporting both 1.0.0 and develop without changing the code when switching.

pnstickne replied:
New additions can't be breaking changes, as they aren't used yet.

That said, whatever the _desired target API_, should be the end goal - with implementation to that starting at additions.

In both cases, only one singular-message form has value being added.

lifflander replied:
Please see comment below.

pnstickne commented on src/vt/vrt/collection/broadcast/broadcastable.impl.h:
```diff
@@ -97,14 +97,59 @@ messaging::PendingSend Broadcastable<ColT,IndexT,BaseProxyT>::broadcast(Args&&..
+}
+
+template <typename ColT, typename IndexT, typename BaseProxyT>
+template <
```
template layout (placement of `>`) is not consistent with header/elsewhere.

pnstickne commented on src/vt/vrt/collection/manager.h:
```diff
@@ -904,6 +904,57 @@ struct CollectionManager
+  template <typename MsgT, typename ColT>
+  messaging::PendingSend broadcastMsgCollectiveImpl(
+    CollectionProxyWrapType<ColT> const& proxy, MsgT* msg,
+    bool instrument = true);
```
As an internal implementation with low call count, require caller to supply value for explicitness.

pnstickne commented on src/vt/vrt/collection/manager.h:
```diff
@@ -904,6 +904,57 @@ struct CollectionManager
+   * \return a pending send
+   */
+  template <typename MsgT, typename ColT>
+  messaging::PendingSend broadcastMsgCollectiveImpl(
```
One could also supply a `MsgPtr<T>&..` here, especially for an internal method which should probably control the lifetime from the call-site.

pnstickne commented on src/vt/vrt/collection/broadcast/broadcastable.impl.h:
```diff
@@ -97,14 +97,61 @@ messaging::PendingSend Broadcastable<ColT,IndexT,BaseProxyT>::broadcast(Args&&..
+template <typename ColT, typename IndexT, typename BaseProxyT>
+template <typename MsgT, ActiveColMemberTypedFnType<MsgT, ColT> f>
+messaging::PendingSend
+Broadcastable<ColT, IndexT, BaseProxyT>::broadcastCollective(MsgSharedPtr<MsgT> msg) const {
```
Likewise here, separate forwarding method and method accepting existing message as well as combine Msg*/MsgPtr.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  14
- tests/unit/collection/test_collection_group.extended.cc  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

pnstickne commented on src/vt/vrt/collection/manager.h:
```diff
@@ -904,6 +904,57 @@ struct CollectionManager
+    typename MsgT,
+    ActiveColTypedFnType<MsgT,typename MsgT::CollectionType> *f
+  >
+  messaging::PendingSend broadcastMsgCollective(
```
`broadcastMsgCollective` vs `broadcastCollective`.. is there a cleaner name to give this for better differentiation? Or is does this improve current consistency?

As other comments, I'm suggesting the form that accept a Msg ends in ..Msg, so that would be `broadcastMsgCollectiveMsg`..

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_collection_group.extended.cc  4
- src/vt/vrt/collection/manager.impl.h  2
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

JacobDomagala commented on src/vt/vrt/collection/manager.impl.h:
```diff
@@ -849,6 +849,82 @@ messaging::PendingSend CollectionManager::broadcastFromRoot(MsgT* raw_msg) {
+
+  auto const cur_epoch = theMsg()->setupEpochMsg(msg);
+
+  return schedule(msg, false, cur_epoch, [proxy, msg] {
```
Not sure if we want to use some conditional check whether it should be scheduled or invoked immediately

lifflander replied:
Let's always schedule it so it hits the queue without question. This will allow priorities to be followed.

JacobDomagala commented:
PR updated:
- addressed formatting issues
- updated API with MsgPtrThief instead of raw/smart msg pointer
- handlers on collection elements are now scheduled instead of being invoked right away

lifflander commented:
@JacobDomagala @pnstickne @PhilMiller @nlslatt 

Please see by comments above. I **do not** want to break the existing `broadcast` API. And, I want `broadcastCollective` to be symmetric with it. We are currently dealing with a lot of issues regarding supporting both 1.0.0 and develop simultaneously for EMPIRE. More breaking API changes, especially to something as fundamental as broadcasts, would not be good.

Do we know of any real cases where overloading to forward the arguments to build the message inline actually causes problems?

lifflander commented:
> @JacobDomagala @pnstickne
> 
> Please see by comments above. I **do not** want to break the existing `broadcast` API. And, I want `broadcastCollective` to be symmetric with it. We are currently dealing with a lot of issues regarding supporting both 1.0.0 and develop simultaneously for EMPIRE. More breaking API changes, especially to something as fundamental as broadcasts, would not be good.
> 
> Do we know of any real cases where overloading to forward the arguments to build the message inline actually causes problems?

@JacobDomagala @nlslatt @PhilMiller @pnstickne  Let's deprecate the overloads for `broadcast`, and `send` that build messages in favor of a new API `broadcastMsg` and `sendMsg`. Then we can use a matching style for `broadcastCollective{,Msg}`.

From reading the EMPIRE code, it uses `broadcast` and `send` to build messages exclusively (I believe). So we shouldn't even get deprecation warnings for most the callsites (unless I missed some).

We need to make this change across all interfaces. This includes `broadcast` and `send` for both collections and **objgroup** proxies.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/manager.impl.h  2
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
- tests/unit/collection/test_collection_group.extended.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

JacobDomagala commented on src/vt/vrt/collection/broadcast/broadcastable.impl.h:
```diff
@@ -94,15 +103,56 @@ template <
+template <typename ColT, typename IndexT, typename BaseProxyT>
+template <typename MsgT, ActiveColMemberTypedFnType<MsgT, ColT> f>
+messaging::PendingSend
+Broadcastable<ColT, IndexT, BaseProxyT>::broadcastMsg(messaging::MsgPtrThief<MsgT> msg) const {
```
I was thinking whether or not to name it `broadcastRootedMsg` so it's more explicit.

lifflander replied:
Let's leave the rooted renaming for the future. We could create the aliases now though so apps can start using that to be more explicit. I'm fine either way.

JacobDomagala commented on src/vt/vrt/collection/broadcast/broadcastable.impl.h:
```diff
@@ -78,7 +78,16 @@ template <
+Broadcastable<ColT, IndexT, BaseProxyT>::broadcastMsg(messaging::MsgPtrThief<MsgT> msg) const {
+  auto proxy = this->getProxy();
+
+  return theCollection()->broadcastMsg<MsgT, f>(proxy, msg.msg_.get());
```
I don't quite like `msg.msg_.get())` here but i don't know if we want to update `CollectionProxyWrapType`'s API to take `MsgThief` instead of raw message ptr.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/manager.impl.h  2
- tests/unit/collection/test_collection_group.extended.cc  4
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
- src/vt/vrt/collection/manager.impl.h  2
- tests/unit/collection/test_collection_group.extended.cc  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/collection/test_collection_group.extended.cc  4
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
- src/vt/vrt/collection/send/sendable.impl.h  6
- src/vt/vrt/collection/manager.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
- tests/unit/collection/test_collection_group.extended.cc  4
- src/vt/vrt/collection/manager.impl.h  2
- src/vt/vrt/collection/send/sendable.impl.h  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
- src/vt/vrt/collection/send/sendable.impl.h  6
- tests/unit/collection/test_collection_group.extended.cc  4
- src/vt/vrt/collection/manager.impl.h  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

JacobDomagala commented on src/vt/vrt/collection/balance/node_stats.cc:
```diff
@@ -240,6 +240,11 @@ getRecvSendDirection(CommKeyType const& comm) {
+  // Comm stats are not recorded for collective bcast
+  // this case is just to avoid warning of not handled enum
+  case CommCategory::CollectiveNodeToCollectionEdges:
+    return std::make_pair(ElementIDType{}, ElementIDType{});
```
Currently we don't record communication part for collective bcast so this is to avoid warning

PhilMiller commented on docs/md/node-stats.md:
```diff
@@ -55,6 +55,7 @@ the code.
 | 4     | `CommCategory::Broadcast` | A broadcast from a collection element to a whole collection (receive-side) |
 | 5     | `CommCategory::CollectionToNodeBcast` | A broadcast from a collection element to all nodes (receive-side) |
 | 6     | `CommCategory::NodeToCollectionBcast` | A broadcast from a node to a whole collection (receive-side) |
+| 7     | `CommCategory::CollectiveNodeToCollectionEdges` | Collective broadcast from a node to all its edges (receive-side) |
```
I think the naming here is a bit off. 
- It's not from "a node", it's from *all* nodes, locally.
- We don't otherwise talk about "edges" as recipients of messages here, we talk about collection elements.

So, I'd suggest describing this
> Collective 'broadcast' from every node to the local collection elements

And naming it maybe `CollectiveToCollectionBcast`?

PhilMiller commented on src/vt/objgroup/proxy/proxy_objgroup.h:
```diff
@@ -102,6 +102,7 @@ struct Proxy {
    * \param[in] msg raw pointer to the message
    */
   template <typename MsgT, ActiveObjType<MsgT, ObjT> fn>
+  [[deprecated("Use broadcastMsg instead")]]
```
Could these deprecation warnings also suggest (maybe even preferentially) switching away from explicit message instantiation to the `broadcast(Args...)` interface?

PhilMiller commented on src/vt/vrt/collection/manager.impl.h:
```diff
@@ -849,6 +849,82 @@ messaging::PendingSend CollectionManager::broadcastFromRoot(MsgT* raw_msg) {
+
+#if vt_check_enabled(lblite)
+  msg->setLBLiteInstrument(instrument);
+  msg->setCat(balance::CommCategory::NodeToCollectionBcast);
```
Isn't this where we wanted to use the new `CommCategory` value?

JacobDomagala replied:
Yes i noticed this also when i was addressing your comment about renaming it, not sure how i missed that.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/manager.impl.h  2
- tests/unit/collection/test_collection_group.extended.cc  4
- src/vt/vrt/collection/send/sendable.impl.h  6
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

PhilMiller commented on tests/unit/collection/test_collection_group.extended.cc:
```diff
@@ -73,18 +82,43 @@ struct ColA : Collection<ColA,Index1D> {
-    EXPECT_TRUE(finished);
+    if (reduce_test) {
+      EXPECT_TRUE(finished);
+    }
```
As I've said on other occasions, I'm rather opposed to the implicit testing that happens by sticking checks in collection element destructors. If `runBcastTestHelper` used `runInEpochCollective(func)`, then we could just do a direct check (i.e. a `foreach` on the collection elements) of `finished`

JacobDomagala replied:
This check in destructor is part of **_test_collection_group_1_** for rooted bcast (which was already implemented before). Tests i wrote for collective bcast don't rely on this check (i use direct check after each bcast call).

I didn't want to change the existing test so i did add `reduce_test` so this check only applies to **_test_collection_group_1_** and not my tests.

lifflander commented on examples/collection/jacobi1d_vt.cc:
```diff
@@ -132,7 +132,7 @@ struct LinearPb1DJacobi : vt::Collection<LinearPb1DJacobi,vt::Index1D> {
       //
       auto proxy = getCollectionProxy();
       auto loopMsg = vt::makeMessage<BlankMsg>();
-      proxy.broadcast<BlankMsg, &LinearPb1DJacobi::sendInfo>(loopMsg.get());
```
I agree with @PhilMiller on these. Can we switch them to the other form and move the arguments down?

lifflander reviewed (approved):
This looks good to me! I agree with @PhilMiller that in the examples we might want to switch to using `broadcast`, moving the arguments down.

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/vt/vrt/collection/send/sendable.impl.h  6
- tests/unit/collection/test_collection_group.extended.cc  4
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
- src/vt/vrt/collection/manager.impl.h  2
         

Clones removed
==============
+ examples/collection/jacobi1d_vt.cc  -1
+ tests/unit/pipe/test_callback_bcast_collection.extended.cc  -4
+ tests/unit/pipe/test_callback_send_collection.extended.cc  -4
+ examples/collection/jacobi2d_vt.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

lifflander commented on examples/collection/jacobi1d_vt.cc:
```diff
@@ -259,14 +258,17 @@ struct LinearPb1DJacobi : vt::Collection<LinearPb1DJacobi,vt::Index1D> {
     if (size_t(myIdx) < numObjs_ - 1) {
-      auto rightMsg = vt::makeMessage<VecMsg>(myIdx, told_[numRowsPerObject_]);
-      proxy[myIdx+1].send<VecMsg, &LinearPb1DJacobi::exchange>(rightMsg.get());
+      auto rightMsg = vt::makeMessage<VecMsg>();
```
Codacy found an issue: [Variable 'rightMsg' is assigned a value that is never used.](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/manager.impl.h  2
- src/vt/vrt/collection/send/sendable.impl.h  6
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
- tests/unit/collection/test_collection_group.extended.cc  4
         

Clones removed
==============
+ examples/collection/jacobi1d_vt.cc  -1
+ tests/unit/pipe/test_callback_send_collection.extended.cc  -4
+ tests/unit/pipe/test_callback_bcast_collection.extended.cc  -4
+ examples/collection/jacobi2d_vt.cc  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)

lifflander commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/vrt/collection/send/sendable.impl.h  6
- tests/unit/collection/test_collection_group.extended.cc  4
- src/vt/vrt/collection/broadcast/broadcastable.impl.h  10
- src/vt/vrt/collection/manager.impl.h  2
         

Clones removed
==============
+ tests/unit/pipe/test_callback_send_collection.extended.cc  -4
+ examples/collection/jacobi1d_vt.cc  -1
+ examples/collection/jacobi2d_vt.cc  -1
+ tests/unit/pipe/test_callback_bcast_collection.extended.cc  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=6206317&bid=20242790)
